File: src/main/java/org/tensorflow/data/Dataset.java
package org.tensorflow.data;

import java.util.Iterator;
import java.util.stream.Stream;

/**
 * Represents a potentially large list of independent elements (samples),
 * and allows iteration and transformations to be performed across
 * these elements.
 */
public interface Dataset<T> {

    /**
     * Creates a new stream of dataset elements (data batches).
     */
//    Stream<T> stream();

    /**
     * Creates a new iterator over dataset elements (data batches).
//     */
//    Iterator<T> iterator();

    /**
     * Groups elements of this dataset into batches.
     *
     * @param batchSize     The number of desired elements per batch
     * @param dropLastBatch Whether to leave out the final batch if
     *                      it has fewer than `batchSize` elements.
     * @return A batched Dataset
     */
    Dataset<T> batch(long batchSize, boolean dropLastBatch);

    default Dataset<T> batch(long batchSize) {

        return batch(batchSize, true);
    }


    long getBatchSize();
    long numBatches();

    long numTensors();
    long numElementsPerTensor();
}

File: src/main/scala/org/tensorflow/data/Dataset.scala
package org.tensorflow.data

trait Dataset[T] extends AutoCloseable {
  /**
    * Combines consecutive elements of this dataset into batches. Does not drop the last batch, even
    * if it has fewer than batchSize elements.
    *
    * @param batchSize The number of consecutive elements of this dataset to combine in a single
    *                  batch.
    * @return A `Dataset`
    */
  def batch(batchSize: Long): Dataset[T] = this.batch(batchSize, dropRemainder = false)

  /**
    * Combines consecutive elements of this dataset into batches.
    *
    * @param batchSize     The number of consecutive elements of this dataset to combine in a single
    *                      batch.
    * @param dropRemainder A boolean representing whether the last batch should be dropped in the
    *                      case that it has fewer than `batchSize` elements.
    * @return A `Dataset`
    */
  def batch(batchSize: Long, dropRemainder: Boolean): Dataset[T]

  def batchSize: Long

  def numBatches: Long

  override def close(): Unit
}
